[ObjCDirectPreconditionThunk] Adding a flag to  with objc_direct symbols' prefix

DataCorrupted opened the pull request:
## TL;DR

This is a stack of PRs implementing features to expose direct methods ABI.
You can see the RFC, design, and discussion [here](https://discourse.llvm.org/t/rfc-optimizing-code-size-of-objc-direct-by-exposing-function-symbols-and-moving-nil-checks-to-thunks/88866).

The stack of the following four PRs completes the whole feature.

https://github.com/llvm/llvm-project/pull/170616 **Flag `-fobjc-direct-precondition-thunk` set up**
https://github.com/llvm/llvm-project/pull/170617 Code refactoring to ease later reviews
https://github.com/llvm/llvm-project/pull/170618 Thunk generation
https://github.com/llvm/llvm-project/pull/170619 Optimizations, some class objects can be known to be realized

## Implementation details

1. Add a flag. I used `-fobjc-direct-precondition-thunk` instead of `-fobjc-direct-caller-thunks` as discussed in this PR.

2. Clean up and set up helper functions to implement later
    a. `canMessageReceiverBeNull` / `canClassObjectBeUnrealized` these two functions will be helpful later to determine which function (true implementation or nil check thunk) we should dispatch a call to. Formatting.
    b. `getSymbolNameForMethod` has a new argument `includePrefixByte`, which allows us to erase the prefixing `\01` when the flag is enabled
    c. `usePreconditionThunk` is the single source of truth of what we should do. It not only checks for the flag, but also whether the method is qualified and we are in the right runtime. A method that `usePreconditionThunk` is either `shouldHavePreconditionThunk` or `shouldHavePreconditionInline`.

## Tests

Driver tests

DataCorrupted committed:
[ExposeObjCDirect] Adding a flag to allow new objc direct ABI

1. Add a flag
2. Clean up and set up helper functions to implement later

Signed-off-by: Peter Rong <PeterRong@meta.com>

DataCorrupted committed:
format

DataCorrupted commented:
TODO: We switch to `-fobjc-direct-caller-thunks` or `-fobjc-direct-nil-check-thunk`. Would love to hear ppl's opinions if any. Naming things are hard.

I will do that after later PRs are merged into this branch, otherwise there will be a lot of conflicts to resolve (later PRs already used the flag in the test)

cc @kyulee-com

tarunprabhu reviewed (changes requested):
I have only reviewed the driver-related code. I cannot comment on anything else.

DataCorrupted committed:
update driver behavior and a test

tarunprabhu reviewed (commented):
Thanks for the changes. The driver-related changes LGTM. 

I have not reviewed the rest of this PR.

rjmccall reviewed (commented):
I'm not sure how I feel about this use of "expose" to specifically mean "don't put the prefix byte on it". What linkage and visibility do these end up getting?

The patch otherwise seems fine.

DataCorrupted committed:
Rename to precondition thunk

DataCorrupted committed:
update comment

kyulee-com commented on clang/lib/CodeGen/CodeGenModule.h:
```diff
@@ -717,6 +717,32 @@ class CodeGenModule : public CodeGenTypeCache {
+  /// Returns false if OMD is null or not a direct method.
+  bool usePreconditionThunk(const ObjCMethodDecl *OMD) const {
+    return OMD && OMD->isDirectMethod() &&
+           getLangOpts().ObjCRuntime.isNeXTFamily() &&
```
Is there a technical reason to restrict this to Apple platforms, or is it just for initial deployment safety? In either case, can you add a comment on it?

DataCorrupted replied:
`objc_direct` is an apple only attribute, I will add a comment.

kyulee-com replied:
https://github.com/llvm/llvm-project/blob/main/clang/test/CodeGenObjC/gnustep2-direct-method.m appears not an apple.

DataCorrupted replied:
https://clang.llvm.org/doxygen/ObjCRuntime_8h_source.html#l00467

I didn't realize that. It appears `clang::ObjCRuntime::allowsDirectDispatch` is the best source of truth whether the runtime supports `objc_direct`; it also specified that GNUStep 2.2+ supports it.

I'll use `allowsDirectDispatch` for clarity. But I'll keep the  `isNeXTFamily` check to only support this feature in NeXT. The rational is that we want to validate the feature real quick but not worrying other runtimes, and migrate to GNU once we are confident.

DataCorrupted committed:
Update clang/include/clang/Options/Options.td

Co-authored-by: Kyungwoo Lee <kyulee@meta.com>

DataCorrupted committed:
update comments

DataCorrupted committed:
comment

DataCorrupted committed:
remove one helper

DataCorrupted committed:
comment and warning about other runtimes

DataCorrupted committed:
Use explict target triples to override host's default

When the driver is tested on Linux or Windows, clang may assume runtime based on host, and fail the tests that originally worked on Apple devices

tarunprabhu reviewed (approved):
The driver-related changes LGTM.
